### DISCUSSION TITLE
Missing password for p12 signing key when root password stored in Mac Keychain

### DISCUSSION THREAD

**Discussion by csandels** [hydraulic-software/conveyor#89] [Q&A] [answered]
The Hydraulic root password is stored in my Mac login keychain.  I can see it in the Keychain Access app as an app "password".  I never set a password for this key but the Conveyor docs imply that I have.  

I tried to set a password for this root key with:
$ conveyor keys passphrase

But the cli tells me "Signing key is stored in your keyring, no need to set up a passphrase".  This implies there is no password on it and the docs clearly state no password is needed since only the conveyor command is authorized to access the key and it's protected by the login password.

However, I already have a signing key from Apple.  It is also in my login keychain, viewable in the Keychain Access app.  Conveyor docs say I should export the public and private version of this key into a p12 file and use the same password for it as was used for the root key.  But no password was used for the root key, and I can't set a password on the root key.

I've set the following in the Conveyor config:

mac.signing-key = "[absolutepath]/AppleCerts.p12"
mac.certificate = "[absolutepath]/AppleCerts.p12"

I run:
$ conveyor make site

I get:
❌ An encrypted private key was found but the password was wrong.

There are no docs about what the password should be, other than Step 3 on "Keys & Certificates" page where it describes steps to export the keys from Keychain Access saying "Step 3. Use the same passphrase as you used for your Conveyor root key. Click OK.".  There is not passphrase on the Conveyor root key.  What is conveyor using to try and decrypt the key?

**Comment by leo-hydraulic** [comment, marked as answer]
Please try running conveyor with flag "--passphrase=[the passphrase you used on your p12 file]".
And thank you for the report, we should improve our documentation in the case where the root key is stored in the keyring, and also fix it to ask for a passphrase in the cases where the mac signing key requires it.

**Reply by leo-hydraulic** [reply]
Alternatively, you can set your config like this:
```hocon
app.mac.signing-key.file.path = "[absolutepath]/AppleCerts.p12"
app.mac.signing-key.file.password = ${env.SOME_ENVIRONMENT_VARIABLE_WITH_YOUR_PASSPHRASE}
```

You don't need to set mac.certificate if it's the same file.

**Reply by csandels** [reply]
Thank you Leo.  This does solve the immediate problem.  However, could you also elaborate on the solution where the root key would be used at all?  The docs and sample are a little conflicting/confusing.

I believe the idea is that we can use
`$ conveyor keys generate`
to generate the defaults.conf file and the two csr files (apple and windows).  By the way, the docs say the conf file will be in our user home directory but actually it's put in the ~/Library/Preferences/Hydraulic/Conveyor.  This command will also put the root key in the login keychain so we can see it in the Keychain Access app.  The defaults.conf file tells us we can configure signing using this:
`app.signing-key = "keyring"`

The note in the config, and the docs, mention when we use the keyring method, the other keys are "derived" from our root key so we don't have to add the signing keys to the config.  The docs also say we should use:
`app.signing-key = "derived"`

So, presumably if we submit the csr generated by "conveyor keys generate" to Apple to get our Developer Id key, which is then imported into Keychain Access (and maybe doesn't need to be), then we don't have to specify the signing key at all in the Conveyor config.  We can just use "keyring" or "derived" (but which one).  But then, how is the signing key found/derived?  Does conveyor just look up the root key and open it by asking the user for the login password?  Then it continues to construct the signing key using the root key and the certificate algorithm it used during csr generation?  And if this is the case... do we need to store the signing cert anywhere locally?

Thank you for your explanation.  It would be great to see some docs explaining the details around this so it's less "magical" in case things don't go as expected.

**Reply by leo-hydraulic** [reply]
Hello, csandels.

> By the way, the docs say the conf file will be in our user home directory but actually it's put in the ~/Library/Preferences/Hydraulic/Conveyor.

Technically, the `~` is within your home directory, but I agree it may sound confusing, I'll clarify it in the docs.

> The docs also say we should use:
`app.signing-key = "derived"`

Sorry, but I haven't been able to find that suggestion anywhere in the docs. Would you mind pointing me to the location in the docs that say that?

> So, presumably if we submit the csr generated by "conveyor keys generate" to Apple to get our Developer Id key, which is then imported into Keychain Access (and maybe doesn't need to be), then we don't have to specify the signing key at all in the Conveyor config. We can just use "keyring" or "derived" (but which one).

You don't need to change anything to use keys derived from your root key.
The value of `app.signing-key`, which should be in your `defaults.conf`, will either be `keyring` indicating the key is stored in the keyring, or the plain value of the root key, which is a list of words.

> But then, how is the signing key found/derived?

It's simply derived algorithmically from the root key.

> Does conveyor just look up the root key and open it by asking the user for the login password?

Usually Conveyor doesn't need to ask for a user password to take the root key, because it originally stores the root key in the keyring. It just has direct access (and no other apps have).

> Then it continues to construct the signing key using the root key and the certificate algorithm it used during csr generation?

Yes, that's roughly how it happens.

>  And if this is the case... do we need to store the signing cert anywhere locally?

Ah, that's a common source of confusion: the certificate is *not* the signing key. Digital code signing is based on a public/private key pair, of which the private key is the signing key, and the public key is stored in the certificate. When someone says "sign with this certificate", what it actually means is "sign with the private key that pairs with the public key stored in this certificate".

Besides the public key of the pair, the certificate contains metadata about you/your company (which is information you send to the CA), and most importantly, the certificate signature, which is another digital signature done by the CA that certifies the rest of the information stored in it.

The certificate itself is public information. You have to store it because it contains information that Conveyor cannot derive.

In your `conveyor.conf` file, just set:

```hocon
app.mac.certificate = path/to/certificate/file.cer
```

pointing it to the file returned by Apple.

I hope this clarifies the confusion. Please let us know if you need assistance with anything else.

**Reply by csandels** [reply]
Thanks for the explanation.  Very helpful.  By the way, the term "derived" is mentioned in config settings on https://conveyor.hydraulic.dev/12.1/configs/#appsigning-key where it explains app.{mac,windows}.signing-key as:
`The private key to use for each platform. If they're set to derived (the default) then derivation from the root...`
"set to derived" seemed to me to be saying when "app.mac.singing-key = derived".

**Reply by leo-hydraulic** [reply]
Ah, yes, that is correct actually.
The root key (`app.signing-key`) is never derived, but the specific signing keys (`app.{mac,windows}.signing-key`) can be derived from the root key, which is indicated by the `derived` special value. But that's an internal detail, you pretty much never have to set it explicitly.
The root key can be stored in the keyring, but the others can't (not supported at the moment).

**Comment by mikehearn** [comment]
The newly released Conveyor 13 handles this situation better.
